arborio/swcio.cpp

```diff
@@ -0,0 +1,503 @@
+        << record.parent_id << '\n';
+
+    out.flags(flags);
+
+    return out;
+}
+
+std::istream& operator>>(std::istream& in, swc_record& record) {
+    std::string line;
+    if (!getline(in, line, '\n')) return in;
+
+    swc_record r;
+    std::istringstream s(line);
+    s >> r.id >> r.tag >> r.x >> r.y >> r.z >> r.r >> r.parent_id;
+    if (s) {
+        record = r;
+    }
+    else {
+        in.setstate(std::ios_base::failbit);
+    }
+
+    return in;
+}
+
+// Parse SWC format data (comments and sequence of SWC records).
+
+static std::vector<swc_record> sort_and_validate_swc(std::vector<swc_record> records, swc_mode mode) {
+    if (records.empty()) return {};
+
+    std::unordered_set<int> seen;
+    std::size_t n_rec = records.size();
+    int first_id = records[0].id;
+    int first_tag = records[0].tag;
+
+    if (records.size()<2) {
+        throw swc_spherical_soma(first_id);
+    }
+
+    for (std::size_t i = 0; i<n_rec; ++i) {
+        swc_record& r = records[i];
+
+        if (r.parent_id>=r.id) {
+            throw swc_record_precedes_parent(r.id);
+        }
+
+        if (!seen.insert(r.id).second) {
+            throw swc_duplicate_record_id(r.id);
+        }
+    }
+
+    std::sort(records.begin(), records.end(), [](const auto& lhs, const auto& rhs) { return lhs.id < rhs.id; });
+    bool first_tag_match = false;
+
+    for (std::size_t i = 0; i<n_rec; ++i) {
+        const swc_record& r = records[i];
+        first_tag_match |= r.parent_id==first_id && r.tag==first_tag;
+
+        if ((i==0 && r.parent_id!=-1) || (i>0 && !seen.count(r.parent_id))) {
+            throw swc_no_such_parent(r.id);
+        }
+    }
+
+    if (mode==swc_mode::strict && !first_tag_match) {
+        throw swc_spherical_soma(first_id);
+    }
+
+    return records;
+}
+
+swc_data parse_swc(std::istream& in, swc_mode mode) {
+    // Collect any initial comments (lines beginning with '#').
+
+    swc_data data;
+    std::string line;
+
+    while (in) {
+        auto c = in.get();
+        if (c=='#') {
+            getline(in, line, '\n');
+            auto from = line.find_first_not_of(" \t");
+            if (from != std::string::npos) {
+                data.metadata.append(line, from);
+            }
+            data.metadata += '\n';
+        }
+        else {
+            in.unget();
+            break;
+        }
+    }
+
+    swc_record r;
+    while (in && in >> r) {
+        data.records.push_back(r);
+    }
+
+    data.records = sort_and_validate_swc(std::move(data.records), mode);
+    return data;
+}
+
+swc_data parse_swc(const std::string& text, swc_mode mode) {
+    std::istringstream is(text);
+    return parse_swc(is, mode);
+}
+
+swc_data parse_swc(std::vector<swc_record> records, swc_mode mode) {
+    swc_data data;
+    data.records = sort_and_validate_swc(std::move(records), mode);
+    return data;
+}
+
+arb::segment_tree as_segment_tree(const std::vector<swc_record>& records) {
```

Comment:
For consistency, maybe this should be called `load_swc_arbor`?

Reply:
I am planning on opening another PR right after this to change this and a few other things. I don't want them to be part of this PR cause it's not gonna show the diff.

Reply:
👍